File: c/day03.c
#include <stdio.h>
#include <limits.h>

#define SIZE 2048


typedef struct Vector
{
	int x;
	int y;
} Vector;

int abs(int a)
{
	return a<0 ? -a : a;
}

int manhattan(Vector *a)
{
	int r;
	r = abs(a->x) + abs(a->y);
	return r;
}

Vector add_vec(Vector *a, Vector *b)
{
	Vector v;
	v.x = a->x + b->x;
	v.y = a->y + b->y;
	return v;
}

Vector sub_vec(Vector *a, Vector *b)
{
	Vector v;
	v.x = a->x - b->x;
	v.y = a->y - b->y;
	return v;
}

Vector mul_vec_int(Vector *a, int b)
{
	Vector v;
	v.x = a->x * b;
	v.y = a->y * b;
	return v;
}

int mul_vec_vec(Vector *a, Vector *b)
{
	int v;
	v = a->x * b->x + a->y * b->y;
	return v;
}

// direction vectors are always orthogonal on the grid
int norm(Vector *a)
{
	if(a->x == 0) {
		return a->y;
	}
	else {
		return a->x;
	}
}

Vector normed(Vector *a)
{
	Vector v;
	v.x = a->x / norm(a);
	v.y = a->y / norm(a);
	return v;
}


typedef struct Edge
{
	Vector origin;
	Vector direction;
} Edge;

Vector intersection(Edge *a, Edge *b)
{
	Vector r;
	int v = 0, w = 0;

	Vector *ao = &(a->origin);
	Vector *bo = &(b->origin);
	Vector *ad = &(a->direction);
	Vector *bd = &(b->direction);
	Vector adn = normed(ad);
	Vector bdn = normed(bd);

	r.x = 0;
	r.y = 0;

	if (mul_vec_vec(&adn, &bdn) != 0) {
		return r;
	}

	if(adn.y == 0) {
		v = (bo->x - ao->x) / adn.x;
		w = (ao->y - bo->y) / bdn.y;
	}
	if(adn.x == 0) {
		v = (bo->y - ao->y) / adn.y;
		w = (ao->x - bo->x) / bdn.x;
	}

	if (v < 0 || w < 0 || v > norm(ad) || w > norm(bd)) {
		return r;
	}

	Vector t = mul_vec_int(&adn, v);
	Vector rr = add_vec(ao, &t);

	return norm(&rr) > 0 ? rr : r;
}



typedef struct Wire
{
	Edge edges[SIZE];
	int lengths[SIZE];
	size_t num;
	int length;
} Wire;


void walk(char c, int d, Wire *wire)
{
	size_t ind;

	ind = wire->num;
	Vector *origin = &(wire->edges[ind].origin);

	if(!ind) {
		origin->x = 0;
		origin->y = 0;
	}
	else {
		Vector *porigin = &(wire->edges[ind-1].origin);
		Vector *pdirection = &(wire->edges[ind-1].direction);
		Vector add = add_vec(porigin, pdirection);
		origin->x = add.x;
		origin->y = add.y;
	}

	Vector *direction = &(wire->edges[ind].direction);
	switch(c) {
		case 'L':
			direction->x = -d;
			direction->y = 0;
			break;
		case 'R':
			direction->x = d;
			direction->y = 0;
			break;
		case 'D':
			direction->x = 0;
			direction->y = -d;
			break;
		case 'U':
			direction->x = 0;
			direction->y = d;
			break;
	}
	wire->lengths[ind] = wire->length;
	wire->length += d;
	wire->num +=  1;
}

int part1(Wire *wires)
{
	int i, j;
	Vector inter;
	int min = INT_MAX;
	int temp;
	Wire *a = wires;
	Wire *b = wires+1;

	for(i=0; i<a->num; i++) {
		for(j=0; j<b->num; j++) {
			//printf("%d %d %d %d\n", i, a->num, a->edges[i].direction.x, a->edges[i].direction.y);
			//printf("%d %d %d %d\n", j, b->num, b->edges[j].direction.x, b->edges[j].direction.y);
			inter = intersection(&(a->edges[i]), &(b->edges[j]));
			temp = manhattan(&inter);
			//printf("temp: %d\n", temp);
			if(temp && temp < min) {
				min = temp;
			}
		}
	}
	return min;
}


int main(int argc, char **argv)
{
	static Wire wires[2];
	int d;
	char c;
	int w = 0;

	Wire *wire = wires;
	wire->length = 0;
	wire->num = 0;
	while(scanf("%c%d,", &c, &d) != EOF) {
		if(c == '\n') {
			if(w++ == 1) {
				break;
			}
			wire += 1;
			wire->length = 0;
			wire->num = 0;
			continue;
		}
		walk(c, d, wire);
	}

	printf("part1: %d\n", part1(wires));
	//printf("part2: %d\n", part2(map));

	return 0;
}

File: python/day03.py
#!/usr/bin/env python3

import fileinput
import itertools

class Vector:
    """
    Class representing a 2D vector. We could use numpy, but that is a bit unneccesary for this
    """

    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def __add__(self, other):
        """
        Add two Vectors
        """
        return Vector(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        """
        Subtract two Vectors
        """
        return Vector(self.x - other.x, self.y - other.y)

    def __eq__(self, other):
        """
        Check whether two Vectors are equal
        """
        return self.x == other.x and self.y == other.y

    def __abs__(self):
        """
        Return the norm of a Vector, as integer
        """
        return int(round((self.x**2 + self.y**2)**0.5))

    def __neg__(self):
        """
        Negate a Vector
        """
        return Vector(-self.x, -self.y)

    def __matmul__(self, other):
        """
        Scalar product of two Vectors
        """
        return self.x * other.x + self.y * other.y

    def __mul__(self, other):
        """
        Element-wise multiplication with a Vector or a scalar
        """
        if isinstance(other, Vector):
            return Vector(self.x * other.x, self.y * other.y)
        return Vector(self.x * other, self.y * other)

    def normed(self):
        """
        Return a normed Vector
        """
        return Vector(self.x // abs(self), self.y // abs(self))

    def manhattan(self, other=None):
        """
        Compute the Manhattan distance to a reference Vector, or the origin if `None`
        """
        if other is None:
            other = Vector(0, 0)
        ref = other - self
        return abs(ref.x) + abs(ref.y)

    @classmethod
    def from_direction(cls, direction, length):
        """
        Create a new Vector from a direction character and a length
        """
        new = cls(*{
            'L': (-length, 0),
            'R': (length, 0),
            'U': (0, length),
            'D': (0, -length)
        }[direction])
        return new

    def __str__(self):
        return f'({self.x}, {self.y})'


class Edge:
    """
    Class representing a single edge of the Wire
    
    Each Edge consists of a origin Vector and a direction Vector
    """

    def __init__(self, origin=Vector(), direction=Vector()):
        self.origin = origin
        self.direction = direction

    def intersection(self, other):
        """
        Checks an Edge for intersection with another Edge
        
        If no intersection is found, return None,
        Else return the intersection point as a Vector
        """
        so = self.origin
        sd = self.direction
        sdn = self.direction.normed()
        oo = other.origin
        od = other.direction
        odn = other.direction.normed()

        if sdn @ odn != 0:
            return None

        if sdn.y == 0: # x, y case
            v = (oo.x - so.x) // sdn.x
            w = (so.y - oo.y) // odn.y
        else: # y, x case
            v = (oo.y - so.y) // sdn.y
            w = (so.x - oo.x) // odn.x

        if v < 0 or w < 0 or v > abs(sd) or w > abs(od):
            return None
        else:
            inter = so + sdn * v
            return inter if abs(inter) else None

    def __str__(self):
        return f'Edge ({self.origin} + {self.direction}'

class Wire:
    """
    Class representing a full Wire.
    
    A Wire consists of multiple Edges
    """

    def __init__(self):
        self.edges = []
        self.lengths = []
        self.length = 0

    def walk(self, direction, length):
        """
        Add a new Edge by walking the wire into a direction by length units
        """
        if not self.edges:
            origin = Vector()
        else:
            origin = self.edges[-1].origin + self.edges[-1].direction
        dirvec = Vector.from_direction(direction, length)
        self.edges.append(Edge(origin, dirvec))
        self.lengths.append(self.length)
        self.length += length

    def __str__(self):
        s = 'Wire: '
        s += ' '.join(str(x) for x in self.edges)
        return s

    def __iter__(self):
        return iter(zip(self.edges, self.lengths))


def part1(wires):
    # this can be written nicely in Python 3.8, using assignment expressions
    min_inter = min(i.intersection(j).manhattan() for (i, _), (j, _) in itertools.product(*wires) if i.intersection(j) is not None)
    return min_inter

def part2(wires):
    # this can be written nicely in Python 3.8, using assignment expressions
    min_length = min(li + lj + abs(i.intersection(j) - i.origin) + abs(i.intersection(j) - j.origin)
                     for (i, li), (j, lj) in itertools.product(*wires) if i.intersection(j) is not None)
    return min_length


def main():
    wires = []
    for line in fileinput.input():
        w = Wire()
        ls = line.split(',')
        for vec in ls:
            vec = vec.strip()
            w.walk(vec[0], int(vec[1:]))
        wires.append(w)

    print('part1: {}'.format(part1(wires)))
    print('part2: {}'.format(part2(wires)))


if __name__ == '__main__':
    main()
